sqlproxyccl: change denylist into an access control list

pjtatlow opened the pull request:
To support an IP Allowlist in the sqlproxy, this change extends
the denylist code to make the Watcher support multiple AccessControllers.
Each AccessController is consulted before allowing a connection through,
and rechecked on any changes to the underlying files.

Part of: [CC-8136](https://cockroachlabs.atlassian.net/browse/CC-8136)

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/98153)
<!-- Reviewable:end -->

pjtatlow commented on pkg/ccl/sqlproxyccl/acl/watcher.go:
```diff
@@ -173,14 +160,32 @@ func (l *listener) Less(than btree.Item) bool {
-	if err := list.Denied(cluster); err != nil {
-		return errors.Wrapf(err, "connection cluster '%v' denied", connection.Cluster)
+func checkListeners(
+	listeners *btree.BTree, timesource timeutil.TimeSource, controllers ...AccessController,
```
Ah my thought here was that by passing them in individually rather than the slice, I would avoid races. But you're right, I would need to make a copy of the slice anyways so I will just do that instead.

pjtatlow commented on pkg/ccl/sqlproxyccl/denylist/file_test.go:
```diff
@@ -1,253 +0,0 @@
-// Copyright 2021 The Cockroach Authors.
```
TIL 🤯

pjtatlow replied:
Had to make two commits to do this, but hopefully this looks better now. Github is still showing those files as deleted, but my local git blame seems to be working correctly.

jaylim-crl replied:
By the way, IIRC, in CRDB land, if we move files like that, we should also make sure that the individual commits are compilable to aid git bisect. i.e. `package denylist` should be renamed to `package acl` in the move commit. Someone else could probably confirm.

pjtatlow replied:
Oof okay, that's going to take a lot more changes then. Let me try again.

pjtatlow commented on pkg/ccl/sqlproxyccl/acl/file.go:
```diff
@@ -0,0 +1,142 @@
+// initiating an access controller with file, we can always update the file with correct content during
+// runtime. We don't want sqlproxy fail to start just because there's something wrong with
+// contents of an access control file.
+// TODO(pj): Does the above still apply to the AllowList? Or was that specific to the DenyList?
```
Do you think failing to load the denylist should have the same behavior as failing to load the allowlist?

pjtatlow replied:
Spoke with Jeff, decided we want the sqlproxy to fail on startup if it has any invalid configurations (allow or deny list).

pjtatlow commented on pkg/ccl/sqlproxyccl/acl/watcher.go:
```diff
@@ -173,14 +160,32 @@ func (l *listener) Less(than btree.Item) bool {
+	connection ConnectionTags, timesource timeutil.TimeSource, controllers ...AccessController,
+) error {
+	for _, c := range controllers {
+		if c != nil {
```
In the case where there is an error reading the files on startup, it will return a `nil` AccessController. But if we're going to fail to start in that case, we can remove this check.

Commit:
sqlproxyccl: move denylist files to acl package

Part of: https://cockroachlabs.atlassian.net/browse/CC-8136

Release note: None

Commit:
sqlproxyccl: change denylist into an access control list

To support an IP Allowlist in the sqlproxy, this change extends
the denylist code to make the Watcher support multiple AccessControllers.
Each AccessController is consulted before allowing a connection through,
and rechecked on any changes to the underlying files.

The sqlproxy will also fail to start if it begins with an invalid allow
or deny list, but if invalid files are written later then it increments
a new error metric so we can be alerted and take action to fix it.

Part of: https://cockroachlabs.atlassian.net/browse/CC-8136

Release note: None

JeffSwenson commented on pkg/ccl/sqlproxyccl/acl/watcher.go:
```diff
@@ -0,0 +1,244 @@
+
+func (w *Watcher) addAccessController(controller AccessController, next chan AccessController) {
+	index := len(w.controllers)
+	w.controllers = append(w.controllers, controller)
```
There is still a small race condition here. It's unlikely to be hit since the channel won't emit a new access controller for five seconds, but we should handle it. addAccessController does not lock controllers and updateAccessController copies the list.

Normally it is safe to assume the lock in the function constructing an object, but that assumption is broken if any of the methods called in the constructor spawns goroutines.

pjtatlow replied:
It took me a second to see it, but you're totally right. Fixed!

jeffswenson reviewed (commented):
This is looking great! Once the race is fixed I'll approve the change.

pjtatlow commented on pkg/ccl/sqlproxyccl/acl/access_control.go:
```diff
@@ -0,0 +1,21 @@
+// Copyright 2021 The Cockroach Authors.
```
Good call!

jeffswenson reviewed (approved):
LGTM
